### PR TITLE
next phase of adjoint plugin

### PR DESCRIPTION
**Pull request by tylerflex**
To try to include in `1.9.0rc2` or `1.9.0`

### `JaxCustomMedium` improvements.
- [ ] point-wise VJP evaluation -> interpolation-aware volume integration.
- [ ]  profile and fix slow speed for large number of pixels.

### `JaxPolySlab` improvements
- [ ] Get `JaxPolySlab` gradient working.
- [ ] Add `.slab_bounds` vjp support.

###  `JaxAbstractMedium` improvements
- [ ] Evaluate `.inside()` for custom medium to support accurate medium VJP for complex shapes.

### Pipeline
- [ ] Refactor from `run_fwd` and `run_adj` to single JVP maker.
- [ ] Add shortcut to run fwd and adj sims in batch if a single adjoint source is detected.
- [ ] Simplify subclassing and conversion from regular to Jax types, if possible.

### Miscellaneous
- [ ] How to render jaxtracer types properly in docs?

### CODE REVIEW

File: tidy3d/components/medium.py [hunk 1]
```diff
@@ -615,6 +615,56 @@ def _interp(
...
+
+            def make_bound_coords(coords: np.ndarray, pt_min: float, pt_max: float) -> List[float]:
+                """Convert user supplied coords into boundary coords to use in :class:`.Grid`."""
+
+                # get coordinates of the bondaries halfway between user-supplied data
+                coord_bounds = (coords[1:] + coords[:1]) / 2.0
+
+                # filter out boundaries that lie outside geometry bounds
+                coord_bounds[coord_bounds <= pt_min] = pt_min
+                coord_bounds[coord_bounds >= pt_max] = pt_max
```
Comment (momchil-flex) [inline tidy3d/components/medium.py:637]: Note that this doesn't technically filter out the coordinates; it truncates all coordinates that are out of bounds to the geometry bounds. So coord_bounds does not change size, but if you use this e.g. in volume integration, all of the elements that were snapped to the geometry boundary will have zero volume element.

File: tidy3d/components/medium.py [hunk 2]
```diff
@@ -615,6 +615,56 @@ def _interp(
...
+        pt_maxs = dict(zip("xyz", rmax))
+
+        def make_grid(scalar_field: ScalarFieldDataArray) -> Grid:
+            """Make a grid for a single dataset."""
+
+            def make_bound_coords(coords: np.ndarray, pt_min: float, pt_max: float) -> List[float]:
+                """Convert user supplied coords into boundary coords to use in :class:`.Grid`."""
+
+                # get coordinates of the bondaries halfway between user-supplied data
+                coord_bounds = (coords[1:] + coords[:1]) / 2.0
```
Comment (momchil-flex) [inline tidy3d/components/medium.py:633]: I don't understand this, it looks like you're taking the cell centers? But then you're passing this to a Grid construction, which normally takes boundaries?

File: tidy3d/plugins/adjoint/components/simulation.py [hunk 3]
```diff
@@ -127,13 +128,12 @@ def _no_overlap(cls, val):
             for j, in_struct_bck in enumerate(in_structs_background):
 
                 # if the contracted geometry intersects with a background structure, raise (overlap)
-                if in_geometry.intersects(in_struct_bck.geometry) and not warning_triggered:
+                if in_geometry.intersects(in_struct_bck.geometry):
```
Comment (momchil-flex) [inline tidy3d/plugins/adjoint/components/simulation.py:131]: I'm not sure if the idea here was for this warning to be triggered for all structures now. However, note that all the warnings I get in the slab notebook if I set `spc=0` still always say spacing between structure 0 and 1, never between 1 and 2 or 2 and 3. I guess have a look into this when you update the notebook as we discussed.

File: tidy3d/plugins/adjoint/components/medium.py [hunk 4]
```diff
@@ -190,14 +192,18 @@ def store_vjp(
 
 @register_pytree_node_class
 class JaxCustomMedium(CustomMedium, JaxObject):
-    """A :class:`.Medium` registered with jax."""
+    """A :class:`.CustomMedium` registered with ``jax``.
+    Note: The gradient calculation assumes uniform field across the pixel.
+    Therefore, the accuracy degrades as the pixel size becomes large
+    with respect to the field variation.
+    """
```
[the remark this answers is not in the document]
Comment (tylerflex) [reply tidy3d/plugins/adjoint/components/medium.py:199]: So just to clarify something here:
* Whatever operations used in the vjp maker for CustomMedium do not need to be backpropped. They are simply for evaluating the integral, which is why we can directly use xarray methods on the field data. 
* When determining the vjp, we need to perform a volume integral, which is currently approximating by interpolation at the position of the data point. However, to do this properly, we need to take into account that the CustomMedium has interpolation options of `nearest` and `linear`. 
    * if `nearest`, each dataset has "influence" over its own rectangular domain, so we can determine the vjp by just performing a volume integral over that domain.
    * if `linear`, the domains of each neighboring pixel are overlapping and we'd need to add a spatially-dependent weight this volume integral. I considered disallowing `linear` in `JaxCustomMedium` but tried it with the current point-wise approximation and it worked fine, so leaving it allowed for now.
* One of my next plans (for 1.9rc3 or 1.10) is to do the volume integral for `nearest`, which should handle the z-constant medium properly if field variation along that dimension is large. 
* Any back propagation through `interp` would need to be defined in `JaxDataArray` but it's a bit complicated so I haven't considered that yet. I figure users could implement their own interpolation within their custom functions using `sel`. The thing is that I'm not 100% sure how xarray / scipy handle interpolation, especially in dimensions > 2 but I also haven't looked into it too much yet due to other priorities. Could be nice to have, but doesn't affect the VJP I think.

[the remark this answers is not in the document]
Comment (tylerflex) [reply tidy3d/plugins/adjoint/components/medium.py:199]: side note: I think the way to define `JaxDataArray.interp(**kwargs)` would be to define a `kwarg` and `self.coords` -dependent mask containing the interpolation weights (in the same shape as `self.values`) and then return the result of `jnp.sum(weights * self.values)`. Jax should be able to handle the rest. I just haven't looked into how to construct those weights.

[the remark this answers is not in the document]
Comment (momchil-flex) [reply tidy3d/plugins/adjoint/components/medium.py:199]: Yeah, that makes sense, when I said "back-propping through the sel/interpolate" I was talking exactly about what's happening in CustomMedium on the forward pass depending on `'nearest'` or `'linear'`.

File: tidy3d/components/medium.py (hunk 1 above)
Comment (tylerflex) [reply tidy3d/components/medium.py:637]: Yes, this is exactly what it's intended to do. The reason is that the vjp must return a Datastructure that is identical in form to the original `self` but with different values. With `0` volume elements for coords out of bounds, we ensure that those data points are not counted in the integral, but still show up in the final returned data structure. I can adjust the docstring as it's a relic of the first iteration where I actually filtered them out before realizing my mistake.

File: tidy3d/plugins/adjoint/components/medium.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (momchil-flex) [reply tidy3d/plugins/adjoint/components/medium.py:199]: Yeah that's how it works, `legume` has a `np.interp` vjp, I think I had the impression that it can be somewhat slow and or memory intensive though, probably even more so for >1D. https://github.com/fancompute/legume/blob/cfa800e8ffabde8f5d621a7d38a77c59a1132780/legume/primitives.py#L226-L250

[the remark this answers is not in the document]
Comment (momchil-flex) [reply tidy3d/plugins/adjoint/components/medium.py:199]: Oh well avoiding an explicit `for` loop may help if it's slow...

File: tidy3d/components/medium.py (hunk 2 above)
Comment (tylerflex) [reply tidy3d/components/medium.py:633]: The `coords` store the centers of the cells in the `eps_dataset`, so the boundaries are the halfway points. Am I missing something?

File: tidy3d/plugins/adjoint/components/simulation.py (hunk 3 above)
Comment (tylerflex) [reply tidy3d/plugins/adjoint/components/simulation.py:131]: seems fixed on my end, as discussed

File: tidy3d/components/medium.py (hunk 1 above)
Comment (momchil-flex) [reply tidy3d/components/medium.py:637]: Oh iiiiii seeee

File: tidy3d/plugins/adjoint/components/medium.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (momchil-flex) [reply tidy3d/plugins/adjoint/components/medium.py:199]: yeah why not

File: tidy3d/components/medium.py (hunk 2 above)
Comment (momchil-flex) [reply tidy3d/components/medium.py:633]: Ok, I see what you mean, I think my confusion here was that you're not making a single grid for all three components, but instead an individual grid for every component, where you assume that the values are defined at the grid center.

The only point here is that in the case of a non-uniform grid, grid centers are the halfway points between grid boundaries, but grid boundaries are not the halfway points between grid centers. Consider these two cells: |--x--|---x---|

Separately, isn't there potentially an issue of not covering the full CustomMedium domain since the array of boundaries has one fewer element than the input array of coordinates (the span is missing half a step on each side)?

Comment (tylerflex) [reply tidy3d/components/medium.py:633]: On the first point about nonuniform, seems I did neglect that, I think it's ok for right now but any suggestions on how to fix it?

On second point, later on I add the bounds for the intersection of the geometry and simulation to the boundary coords list, so it takes care of that missing step on each side.

Comment (momchil-flex) [reply tidy3d/components/medium.py:633]: > On the first point about nonuniform, seems I did neglect that, I think it's ok for right now but any suggestions on how to fix it?

Not really, I think there's a piece of information missing. You need to know the size of one of the cells and then the rest are uniquely defined in a domino effect.
